src/parser/string.rs

```diff
@@ -0,0 +1,276 @@
+    must_escape: &'static str,
+) -> impl Fn(Span<'a>) -> IResult<Span<'a>, Span<'a>, ParseError<Span<'a>, F>> {
+    move |from: Span<'a>| {
+        let mut s = String::from(must_escape);
+        s.push(delim);
+        s.push('\\');
+        if !whitespace {
+            for c in crate::parser::LURK_WHITESPACE {
+                s.push(c);
+            }
+        }
+        let mut p = verify(is_not(&*s), |s: &Span<'a>| !s.fragment().is_empty());
+        p(from)
+    }
+}
+
+/// A string fragment contains a fragment of a string being parsed: either
+/// a non-empty Literal (a series of non-escaped characters), a single
+/// parsed escaped character, or a block of escaped whitespace.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum StringFragment<'a> {
+    Literal(Span<'a>),
+    EscapedChar(char),
+    EscapedWS,
+}
+
+/// Combine parse_literal, parse_escaped_whitespace, and parse_escaped_char
+/// into a StringFragment.
+pub fn parse_fragment<'a, F: LurkField>(
+    delim: char,
+    whitespace: bool,
+    must_escape: &'static str,
+) -> impl Fn(Span<'a>) -> IResult<Span<'a>, StringFragment<'a>, ParseError<Span<'a>, F>> {
+    move |from: Span<'a>| {
+        if whitespace {
+            alt((
+                map(
+                    parse_literal(delim, whitespace, must_escape),
+                    StringFragment::Literal,
+                ),
+                map(
+                    parse_escaped_char(delim, must_escape),
+                    StringFragment::EscapedChar,
+                ),
+                value(StringFragment::EscapedWS, parse_escaped_whitespace()),
+            ))(from)
+        } else {
+            alt((
+                map(
+                    parse_literal(delim, whitespace, must_escape),
+                    StringFragment::Literal,
+                ),
+                map(
+                    parse_escaped_char(delim, must_escape),
+                    StringFragment::EscapedChar,
+                ),
+            ))(from)
+        }
+    }
+}
+
+/// Parse a string. Use a loop of parse_fragment and push all of the fragments
+/// into an output string.
+pub fn parse_string_inner1<'a, F: LurkField>(
+    delim: char,
+    whitespace: bool,
+    must_escape: &'static str,
+) -> impl Fn(Span<'a>) -> IResult<Span<'a>, String, ParseError<Span<'a>, F>> {
+    move |from: Span<'a>| {
+        fold_many1(
+            parse_fragment(delim, whitespace, must_escape),
+            String::new,
+            |mut string, fragment| {
+                match fragment {
+                    StringFragment::Literal(s) => string.push_str(s.fragment()),
+                    StringFragment::EscapedChar(c) => string.push(c),
+                    StringFragment::EscapedWS => {}
+                }
+                string
+            },
+        )(from)
+    }
+}
+
+/// Parse a string. Use a loop of parse_fragment and push all of the fragments
+/// into an output string.
+pub fn parse_string_inner<'a, F: LurkField>(
```

Comment:
Those two are tedious to differentiate. If it’s impossible to refactor them to share common code, it would be great to explicitly mention why in a code comment on each function that makes the difference with the other one explicit.

Reply:
I will update  the comment to clarify that `parse_string_inner1` must be a non-empty string. 
As far as sharing code, I think that's unlikely, considering that `nom::fold_many0` https://docs.rs/nom/latest/nom/multi/fn.fold_many0.html and `nom::fold_many1` https://docs.rs/nom/latest/nom/multi/fn.fold_many1.html are implemented separately (albeit with some boilerplate).